File: mesh.js
function DrawMesh(gl)
{
	if (this.vb)
	{
		gl.bindBuffer(gl.ARRAY_BUFFER, this.vb);
		
		var format = this.vertFormat;
		var elements = format.length / 2;
		var offset = 0;
		for (var k = 0; k < elements; k++)
		{
			var idx = k * 2;
			var fmtelement = format.substring(idx, idx + 2);
			switch (fmtelement)
			{
				case "P3":
					gl.enableVertexAttribArray(gl.shaderPositionLocation);
					gl.vertexAttribPointer(gl.shaderPositionLocation, 3, gl.FLOAT, false, this.vertexStride, offset);
					offset += 12;
					break;
				case "N3": 
					if (gl.shaderNormalLocation >= 0)
					{
						gl.enableVertexAttribArray(gl.shaderNormalLocation);
						gl.vertexAttribPointer(gl.shaderNormalLocation, 3, gl.FLOAT, false, this.vertexStride, offset);
					}
					offset += 12;
					break;
				case "T2":
					if (gl.shaderUVLocattion >= 0)
					{
						gl.enableVertexAttribArray(gl.shaderUVLocattion);
						gl.vertexAttribPointer(gl.shaderUVLocattion, 2, gl.FLOAT, false, this.vertexStride, offset);
					}
					offset += 8;
					break;
				default:
					alert("unsupported vertex format: " + fmtelement);
					break;
			}
		}

		if (this.ib)
		{
			// Indexed draw
			gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, this.ib);
			gl.drawElements(gl.TRIANGLES, this.triangleCount * 3, gl.UNSIGNED_SHORT, 0);
		}
		else
		{
			// Non indexed draw
			gl.drawArrays(gl.TRIANGLES, 0, this.triangleCount * 3);
		}
	}
}

function MeshToString()
{
	var str;

	str = "mesh:" + this.name;
	str += ";";
	str += "src:" + this.src;

	return str;
}

function Mesh(scene, name, src)
{
	//this.scene = scene;
	this.name = name;
	this.src = src;

	this.draw = DrawMesh;
	this.toString = MeshToString;
	this.save = SaveMesh;

	this.vb = null;
	this.ib = null;
	this.triangleCount = 0;
	this.vertexStride = 0;
	this.boxMin = vec3.create();
	this.boxMax = vec3.create();

	if (EditorName)
		this.fullPath = GetFullPath(src);

	meshXML = LoadXML(src);
	if( meshXML )
	{
		this.triangleCount = parseInt(meshXML.documentElement.attributes.getNamedItem("triangleCount").value);

		meshChildren = meshXML.documentElement.childNodes;
		for (var i = 0; i < meshChildren.length; i++)
		{
			if (meshChildren[i].nodeType == 1)
			{
				if (meshChildren[i].nodeName == "verts")
				{
					this.vertCount = parseInt(meshChildren[i].attributes.getNamedItem("count").value);
					this.vertFormat = meshChildren[i].attributes.getNamedItem("format").value;


					var posArray = [];
					var nrmArray = [];
					var uvArray = [];

					vertPieces = meshChildren[i].childNodes;
					for (var j = 0; j < vertPieces.length; j++)
					{
						if (vertPieces[j].nodeType == 1)
						{
							if (vertPieces[j].nodeName == "positions")
							{
								var positions = vertPieces[j].childNodes;
								for (var k = 0; k < positions.length; k++)
								{
									if (positions[k].nodeType == 1)
									{
										if (positions[k].nodeName == "v3")
										{
											var posVals = positions[k].textContent.csvToArray();
											var pos = vec3.fromValues(posVals[0][0], posVals[0][1], posVals[0][2]);
											posArray.push(pos);

											vec3.max(this.boxMax, this.boxMax, pos);
											vec3.min(this.boxMin, this.boxMin, pos);
										}
									}
								}
							}
							else if (vertPieces[j].nodeName == "normals")
							{
								var normals = vertPieces[j].childNodes;
								for (var k = 0; k < normals.length; k++)
								{
									if (normals[k].nodeType == 1)
									{
										if (normals[k].nodeName == "v3")
										{
											var vals = normals[k].textContent.csvToArray();
											var nrm = vec3.fromValues(vals[0][0], vals[0][1], vals[0][2]);
											nrmArray.push(nrm);
										}
									}
								}
							}
							else if (vertPieces[j].nodeName == "texcoords")
							{
								var uvs = vertPieces[j].childNodes;
								for (var k = 0; k < uvs.length; k++)
								{
									if (uvs[k].nodeType == 1)
									{
										if (uvs[k].nodeName == "v2")
										{
											var vals = uvs[k].textContent.csvToArray();
											var uv = vec2.fromValues(vals[0][0], vals[0][1]);
											uvArray.push(uv);
										}
									}
								}
							}
						}
					}

					var interleaved = [];

					for (var j = 0; j < this.vertCount; j++)
					{
						var format = this.vertFormat;
						var elements = format.length / 2;
						this.vertexStride = 0;
						for (var k = 0; k < elements; k++)
						{
							var idx = k * 2;
							var fmtelement = format.substring(idx, idx + 2);
							switch (fmtelement)
							{
								case "P3":
									interleaved.push(posArray[j][0]);
									interleaved.push(posArray[j][1]);
									interleaved.push(posArray[j][2]);
									this.vertexStride += 12;
									break;
								case "N3":
									interleaved.push(nrmArray[j][0]);
									interleaved.push(nrmArray[j][1]);
									interleaved.push(nrmArray[j][2]);
									this.vertexStride += 12;
									break;
								case "T2":
									interleaved.push(uvArray[j][0]);
									interleaved.push(uvArray[j][1]);
									this.vertexStride += 8;
									break;
								default:
									alert("unsupported vertex format: " + fmtelement);
									break;
							}
						}
					}

					this.vb = scene.gl.createBuffer();
					scene.gl.bindBuffer(scene.gl.ARRAY_BUFFER, this.vb);
					scene.gl.bufferData(scene.gl.ARRAY_BUFFER, new Float32Array(interleaved), scene.gl.STATIC_DRAW);
				}
				else if (meshChildren[i].nodeName == "indices")
				{
					var vals = meshChildren[i].textContent.csvToArray();

					this.ib = scene.gl.createBuffer();
					scene.gl.bindBuffer(scene.gl.ELEMENT_ARRAY_BUFFER, this.ib);
					scene.gl.bufferData(scene.gl.ELEMENT_ARRAY_BUFFER, new Uint16Array(vals[0]), scene.gl.STATIC_DRAW);
				}
			}
		}
	}
}

function SaveMesh(path)
{
	if (EditorName)
	{
		// Load the original mesh file and save it in the new directory
		var file = LoadFile(this.fullPath);
		SaveFile(path + this.src, file);
	}
}

File: WebGLEditor/Mesh.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;
using OpenTK;
using OpenTK.Graphics.OpenGL;

namespace WebGLEditor
{
    public class Mesh : Asset
    {
        public int vb = 0;
        public int ib = 0;
        public int triangleCount = 0;
        public int vertCount = 0;
        public string vertFormat;

        public Mesh(Scene scene, string name, string src)
            : base(scene, name, src)
        {

            try
            {
                List<Vector3> posArray = new List<Vector3>();
                List<Vector3> nrmArray = new List<Vector3>();
                List<Vector2> uvArray = new List<Vector2>();

                XmlDocument meshXML = new XmlDocument();
                meshXML.Load(src);
		        triangleCount = Convert.ToInt32(meshXML.DocumentElement.Attributes.GetNamedItem("triangleCount").Value);
                
                foreach( XmlNode child in meshXML.DocumentElement.ChildNodes )
		        {
			        if (child.NodeType == XmlNodeType.Element)
			        {
				        if (child.Name == "verts")
				        {
					        vertCount = Convert.ToInt32(child.Attributes.GetNamedItem("count").Value);
					        vertFormat = child.Attributes.GetNamedItem("format").Value;


                            foreach( XmlNode vertPiece in child.ChildNodes )
                            {
						        if (vertPiece.NodeType == XmlNodeType.Element)
						        {
							        if (vertPiece.Name == "positions")
							        {
                                        foreach( XmlNode position in vertPiece.ChildNodes )
								        {
									        if (position.NodeType == XmlNodeType.Element)
									        {
										        if (position.Name == "v3")
										        {
											        string[] posVals = position.InnerText.Split(',');
											        Vector3 pos = new Vector3(Convert.ToSingle(posVals[0]), Convert.ToSingle(posVals[1]), Convert.ToSingle(posVals[2]));
											        posArray.Add(pos);
										        }
									        }
								        }
							        }
							        else if (vertPiece.Name == "normals")
							        {
                                        foreach( XmlNode normal in vertPiece.ChildNodes )
								        {
									        if (normal.NodeType == XmlNodeType.Element)
									        {
										        if (normal.Name == "v3")
										        {
											        string[] vals = normal.InnerText.Split(',');
											        Vector3 nrm = new Vector3(Convert.ToSingle(vals[0]), Convert.ToSingle(vals[1]), Convert.ToSingle(vals[2]));
											        nrmArray.Add(nrm);
										        }
									        }
								        }
							        }
							        else if (vertPiece.Name == "texcoords")
							        {
                                        foreach( XmlNode uvs in vertPiece.ChildNodes )
								        {
									        if (uvs.NodeType == XmlNodeType.Element)
									        {
										        if (uvs.Name == "v2")
										        {
											        string[] vals = uvs.InnerText.Split(',');
											        Vector2 uv = new Vector2(Convert.ToSingle(vals[0]), Convert.ToSingle(vals[1]));
											        uvArray.Add(uv);
										        }
									        }
								        }
							        }
						        }
					        }

                            List<float> interleaved = new List<float>();
                            int vertexSize = 0;
                            switch (this.vertFormat)
                            {
                                case "P3":
                                    for (var j = 0; j < this.vertCount; j++)
                                    {
                                        interleaved.Add(posArray[j].X);
                                        interleaved.Add(posArray[j].Y);
                                        interleaved.Add(posArray[j].Z);
                                    }
                                    vertexSize = 12;
                                    break;
                                case "P3N3":
                                    for (var j = 0; j < this.vertCount; j++)
                                    {
                                        interleaved.Add(posArray[j].X);
                                        interleaved.Add(posArray[j].Y);
                                        interleaved.Add(posArray[j].Z);
                                        interleaved.Add(nrmArray[j].X);
                                        interleaved.Add(nrmArray[j].Y);
                                        interleaved.Add(nrmArray[j].Z);
                                    }
                                    vertexSize = 24;
                                    break;
                                case "P3T2":
                                    for (var j = 0; j < this.vertCount; j++)
                                    {
                                        interleaved.Add(posArray[j].X);
                                        interleaved.Add(posArray[j].Y);
                                        interleaved.Add(posArray[j].Z);
                                        interleaved.Add(uvArray[j].X);
                                        interleaved.Add(uvArray[j].Y);
                                    }
                                    vertexSize = 20;
                                    break;
                                default:
                                    System.Windows.Forms.MessageBox.Show("unsupported vertex format: " + this.vertFormat);
                                    break;
                            }

                            GL.GenBuffers(1, out vb);
                            GL.BindBuffer(BufferTarget.ArrayBuffer, vb);
                            GL.BufferData(BufferTarget.ArrayBuffer, (IntPtr)(interleaved.Count * vertexSize), interleaved.ToArray(), BufferUsageHint.StaticDraw);
				        }
				        else if (child.Name == "indices")
				        {
					        string[] vals = child.InnerText.Split(',');
                            List<UInt16> indices = new List<UInt16>();
                            foreach( string val in vals )
                            {
                                indices.Add(Convert.ToUInt16(val));
                            }
                            
                            GL.GenBuffers(1, out ib);
                            GL.BindBuffer(BufferTarget.ElementArrayBuffer, ib);
                            GL.BufferData(BufferTarget.ElementArrayBuffer, (IntPtr)(indices.Count * 2), indices.ToArray(), BufferUsageHint.StaticDraw);
				        }
			        }
		        }

                /*
                if (posArray.Count > 0)
                {
                    GL.GenBuffers(1, out posBuffer);
                    GL.BindBuffer(BufferTarget.ArrayBuffer, posBuffer);
                    GL.BufferData<Vector3>(BufferTarget.ArrayBuffer, new IntPtr(posArray.Count * Vector3.SizeInBytes), posArray.ToArray(), BufferUsageHint.StaticDraw);
                }
                if (nrmArray.Count > 0)
                {
                    GL.GenBuffers(1, out nrmBuffer);
                    GL.BindBuffer(BufferTarget.ArrayBuffer, nrmBuffer);
                    GL.BufferData<Vector3>(BufferTarget.ArrayBuffer, new IntPtr(nrmArray.Count * Vector3.SizeInBytes), nrmArray.ToArray(), BufferUsageHint.StaticDraw);
                }
                if (uvArray.Count > 0)
                {
                    GL.GenBuffers(1, out uvBuffer);
                    GL.BindBuffer(BufferTarget.ArrayBuffer, uvBuffer);
                    GL.BufferData<Vector2>(BufferTarget.ArrayBuffer, new IntPtr(uvArray.Count * Vector2.SizeInBytes), uvArray.ToArray(), BufferUsageHint.StaticDraw);
                }

                GL.GenVertexArrays(1, out vb);
                GL.BindVertexArray(vb);

                GL.EnableVertexAttribArray(0);
                GL.BindBuffer(BufferTarget.ArrayBuffer, posBuffer);
                GL.VertexAttribPointer(0, 3, VertexAttribPointerType.Float, false, Vector3.SizeInBytes, 0);

                int stream = 1;
                if (nrmBuffer > 0)
                {
                    GL.EnableVertexAttribArray(stream);
                    GL.BindBuffer(BufferTarget.ArrayBuffer, nrmBuffer);
                    GL.VertexAttribPointer(0, 3, VertexAttribPointerType.Float, true, Vector3.SizeInBytes, 0);
                    stream++;
                }
                if (uvBuffer > 0)
                {
                    GL.EnableVertexAttribArray(stream);
                    GL.BindBuffer(BufferTarget.ArrayBuffer, uvBuffer);
                    GL.VertexAttribPointer(0, 2, VertexAttribPointerType.Float, false, Vector2.SizeInBytes, 0);
                    stream++;
                }

                GL.BindBuffer(BufferTarget.ElementArrayBuffer, ib);
                GL.BindVertexArray(0);
                 * */

	        }
            catch(Exception)
            {
                System.Windows.Forms.MessageBox.Show("Failed to load mesh: " + src);
            }
        }

        public void Draw(GLContext gl)
        {
	        if (vb != 0)
	        {
                GL.BindBuffer(BufferTarget.ArrayBuffer, vb);
                
		        switch (vertFormat)
		        {
			        case "P3":
                        GL.EnableVertexAttribArray(gl.shaderPositionLocation);
                        GL.VertexAttribPointer(gl.shaderPositionLocation, 3, VertexAttribPointerType.Float, false, 0, 0);
				        break;
			        case "P3N3":
				        GL.EnableVertexAttribArray(gl.shaderPositionLocation);
				        GL.VertexAttribPointer(gl.shaderPositionLocation, 3, VertexAttribPointerType.Float, false, 24, 0);

				        if( gl.shaderNormalLocation >= 0 )
				        {
					        GL.EnableVertexAttribArray(gl.shaderNormalLocation);
					        GL.VertexAttribPointer(gl.shaderNormalLocation, 3, VertexAttribPointerType.Float, false, 24, 12);
				        }
				        break;
			        case "P3T2":
				        GL.EnableVertexAttribArray(gl.shaderPositionLocation);
				        GL.VertexAttribPointer(gl.shaderPositionLocation, 3, VertexAttribPointerType.Float, false, 20, 0);

				        if( gl.shaderUVLocattion >= 0 )
				        {
					        GL.EnableVertexAttribArray(gl.shaderUVLocattion);
					        GL.VertexAttribPointer(gl.shaderUVLocattion, 2, VertexAttribPointerType.Float, false, 20, 12);
				        }
				        break;
			        default:
				        System.Windows.Forms.MessageBox.Show("unsupported vertex format: " + vertFormat);
				        break; 
		        }
                //GL.BindVertexArray(vb);

		        if (ib > 0)
		        {
			        // Indexed draw
                    GL.BindBuffer(BufferTarget.ElementArrayBuffer, ib);
                    GL.DrawElements(BeginMode.Triangles, triangleCount * 3, DrawElementsType.UnsignedShort, 0);
                    GL.BindBuffer(BufferTarget.ElementArrayBuffer, 0);
		        }
		        else
		        {
			        // Non indexed draw
                    GL.DrawArrays(BeginMode.Triangles, 0, triangleCount * 3);                    
		        }
                
                switch (vertFormat)
                {
                    case "P3":
                        GL.DisableVertexAttribArray(gl.shaderPositionLocation);
                        break;
                    case "P3N3":
                        GL.DisableVertexAttribArray(gl.shaderPositionLocation);

                        if (gl.shaderNormalLocation >= 0)
                            GL.DisableVertexAttribArray(gl.shaderNormalLocation);
                        break;
                    case "P3T2":
                        GL.DisableVertexAttribArray(gl.shaderPositionLocation);
                        if (gl.shaderUVLocattion >= 0)
                            GL.DisableVertexAttribArray(gl.shaderUVLocattion);
                        break;
                    default:
                        System.Windows.Forms.MessageBox.Show("unsupported vertex format: " + vertFormat);
                        break;
                }
                GL.BindBuffer(BufferTarget.ArrayBuffer, 0);
	        }
        }
    }
}
